### ISSUE TITLE
[FEA] Multiple threads shareing the same GPU

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#15]
**Is your feature request related to a problem? Please describe.**
Currently all of cudf processing uses a single stream which blocks all other streams from sharing the GPU.  The plugin uses a semaphore to prevent too many processes on the GPU at once.  This works to overlap some I/O with processing, but it does not work if you want to overlap processing.

**Describe the solution you'd like**
I would be great to be able to support multiple streams, possibly using per-thread-default streams.

**Comment by rongou** [comment]
Here is the design proposal.

## Motivation
Spark is multi-threaded on the host (CPU) side. Input is divided into partitions, and multiple tasks are launched on each CPU to work on these partitions. Each task typically runs in a single thread, obtained from a thread pool. This task level parallelism ensures all the CPU cores are utilized to increase overall throughput.

With the `spark-rapids` plugin, some or all of the work in each task is offloaded to the GPU. To efficiently use the GPU, the CUDA operations in these tasks should execute concurrently. CUDA applications manage concurrency by executing asynchronous commands in streams, sequences of commands that execute in order. Currently all host threads share the single default CUDA stream. Also known as the *legacy* default stream or the *NULL* stream, it implicitly synchronizes with itself and all other streams on the device, causing CUDA operations to be serialized.

To increase GPU throughput, it is desirable to execute each task or host thread in a separate, asynchronous stream. Ideally the legacy default stream is never used so as not to trigger implicit synchronization.

## Goals
* Each Spark task or host thread runs in a separate CUDA stream.
* The legacy default stream is not used.
* Test and fix the current code base so that concurrent execution on the GPU works correctly, e.g. no race conditions, deadlocks etc.

## Non-Goals
* Once tasks execute concurrently on GPUs, there may be further opportunities to tune current approaches to get better performance. This is outside the scope of this issue.
* Runtime compilation using NVRTC/Jitify is only for device code, which is oblivious to streams (unless using CUDA dynamic parallelism).

## Assumptions
* There shouldn’t be any big structural changes to the code base, either the underlying C++ libraries, or the Java/Scala code in the `spark-rapids` plugin.
* A Spark task shouldn’t need to manipulate CUDA streams directly, or deal with multiple streams.
* Some background tasks such as asynchronous spill may operate with their own streams.
* GPU memories allocated in one task/stream may be deallocated in another task/stream.

## Risks
* Concurrent execution of GPU code may be hard to reason about or debug. We need good test coverage to make sure the code works correctly.
* Some third party libraries may never be compiled with per-thread default stream enabled. In the case of XGBoost or other machine learning libraries, since model training does not overlap with ETL, using the legacy default stream during training should not have an adverse effect. However, libraries used during ETL may cause excessive synchronization when PTDS is not enabled.
* There are upcoming CUDA features that may impact some key components in the stack.

## Design
CUDA 7 introduced the *per-thread default stream* (PTDS), which has two effects. First, it gives each host thread its own default stream. This means that commands issued to the default stream by different host threads can run concurrently. Second, these default streams are regular streams. This means that commands in the default stream may run concurrently with commands in non-default streams.

To enable PTDS, we can compile with the `nvcc` command-line option `--default-stream per-thread`. Once this is done on all CUDA code, each Spark thread would use its own default stream, and operations from different threads would execute concurrently on the GPU.

![ptds](https://user-images.githubusercontent.com/497101/86392547-7a17b100-bc50-11ea-8a8d-253860100257.png)

The work involved can be roughly divided into two parts, basic mechanism and correctness.
### Basic Mechanism
* [x] Option to enable PTDS for RMM (https://github.com/rapidsai/rmm/pull/354)
* [x] Option to enable PTDS for cuDF ([PR](https://github.com/rapidsai/cudf/pull/4995))
* [x] Option to enable PTDS for cuDF JNI ([PR](https://github.com/rapidsai/cudf/pull/4995))
* [x] Ensure the Thrust library, which is heavily used by cuDF, handles PTDS correctly (https://github.com/thrust/thrust/pull/1128)
* [x] RMM should use the correct version of Thrust (https://github.com/rapidsai/rmm/pull/378)
* [x] cuDF should use the correct version of Thrust (https://github.com/rapidsai/cudf/pull/5315)
* [x] cuDF JNI should use the correct version of Thrust (https://github.com/rapidsai/cudf/pull/5414)

### Correctness
* [ ] Test PTDS in Thrust (https://github.com/thrust/thrust/issues/1132, https://github.com/thrust/thrust/pull/1201)
* [ ] Support PTDS in RMM (https://github.com/rapidsai/rmm/issues/407, https://github.com/rapidsai/rmm/issues/410, https://github.com/rapidsai/rmm/pull/425)
* [ ] Test PTDS in cuDF (https://github.com/rapidsai/cudf/issues/5614)
* [ ] Test PTDS in cuDF JNI

Finally, we should run some benchmarks with PTDS enabled. These can be standard TPC-H or TPC-DS queries, or real customer queries if possible.
* [ ] Benchmarks

After verifying that everything works correctly, the spark-rapids distribution of RMM/cuDF can enable PTDS by default.
* [ ] Enable PTDS in `spark-rapids` distribution

## Alternatives Considered
Most (if not all) cuDF functions support passing in an extra parameter for the stream to execute on, although currently these stream parameters are not exposed in cuDF’s public API. One alternative is to make these parameters public and for the spark-rapids plugin to manage CUDA streams explicitly. Each task can grab its own stream, preferably from a stream pool, and pass it to all subsequent cuDF calls. The benefit of this approach is that the intent is more explicit, and the code doesn’t need to be compiled with the per-thread default stream flag. The drawback is that the client code needs to manage streams, leading to more complexity.

**Comment by jrhemstad** [comment]
> * Some third party libraries may never be compiled with per-thread default stream enabled.

fyi, you can explicitly pass `cudaStreamPerThread` into external libraries to get PTDS behavior even if the external library wasn't complied with PTDS enabled.

**Comment by rongou** [comment]
Closing this as the list of items is done. Will open another one for performance tuning.
